File: services/gateways/bff/src/modules/interview/usecases/cancel-interview.usecase.ts
import { Injectable } from "@nestjs/common";
import { InterviewGrpcService } from "../../../infra/grpc/services/interview-grpc.service";

export class CancelInterviewCommand {
    constructor(
        public readonly interviewId: string,
        public readonly reason?: string,
    ) {}
}

@Injectable()
export class CancelInterviewUseCase {
    constructor(private readonly interviewGrpcService: InterviewGrpcService) {}

    async execute(command: CancelInterviewCommand) {
        const response = await this.interviewGrpcService.cancelInterview(
            command.interviewId,
            command.reason,
        );

        return {
            interviewId: response.interviewId,
            status: response.status,
            endedAt: response.endedAt,
        };
    }
}

File: services/domains/interview/src/main/java/me/unbrdn/core/interview/application/port/in/CancelInterviewUseCase.java
package me.unbrdn.core.interview.application.port.in;

import java.util.UUID;

/** Use Case for cancelling an interview */
public interface CancelInterviewUseCase {
    CancelInterviewResult execute(CancelInterviewCommand command);

    record CancelInterviewCommand(UUID interviewId, String reason) {}

    record CancelInterviewResult(String interviewId, String status, String endedAt) {}
}
